FEAT: Matrix determinant

cemoody opened the pull request:
Implemented a batch matrix determinant function. Requires the matrix inversion in #608 since the gradient of a matrix determinant involves a matrix inversion.  So PR #608 should be merged before this one. Once we have determinants & inverses, should be easier to implement #575

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+        a_type = _convert_type(a_type)
+        utils.type_check.expect(a_type.dtype.kind == 'f')
+        # Only a minibatch of 2D array shapes allowed
+        _check_ndim(a_type, lower=3, upper=3)
```
`type_check.expect(in_type.ndim == 3)` is simpler

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+
+from chainer import cuda
+from chainer import function
+from chainer.functions.array.reshape import Reshape
```
[OpenStack Style Guideline](http://docs.openstack.org/developer/hacking/) which Chainer use as a coding style, forbids import of objects. So please write `from Chainer.functions.array import reshape` .

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+from chainer.functions.math.matmul import _check_ndim
+from chainer.functions.math.matmul import _convert_type
+from chainer.functions.math.matmul import _get_ld
+from chainer.functions.math.matmul import _mat_ptrs
```
According to pep8, objects which begin with single underscore are for private use (see e.g. [here] http://stackoverflow.com/questions/1301346/the-meaning-of-a-single-and-a-double-underscore-before-an-object-name-in-python). So consider make these functions public.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+from chainer import utils
+
+
+def _det_gpu(b):
```
Do you think we should add this function to CuPy, although NumPy does not have corresponding method?

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+
+    def backward_gpu(self, x, gy):
+        # For matrix x, z, y the gradient is
+        # d[det(x)]/dz = gy det(x) (x^-1).T
```
What happens when `x` is not invertible?

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+        return self.detx,
+
+    def backward_gpu(self, x, gy):
+        # For matrix x, z, y the gradient is
```
I am afraid I could not understand the relationship the usage of symbols in this line and that of the next line.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+                _inv_gpu(x.transpose([0, 2, 1])))
+        return utils.force_array(grad),
+
+    def backward_cpu(self, x, gy):
```
I think `backward_cpu` should precede `backward_gpu`, although it just a matter of tastes.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,106 @@
+        a (Variable): Input array to compute the determinant for.
+
+    Returns:
+        ~chainer.Variable: Scalar determinant of the matrix a.
```
Return value of `det` is `xp.array` whose shape is `(1, )`, not a scalar.

```
In [16]: a = chainer.Variable(numpy.random.uniform(-1, 1, (3, 3)).astype(numpy.float32))
In [17]: chainer.functions.det(a).data
Out[17]: array([-0.80874199], dtype=float32)
```

How about extract the value of returned array?

delta2323 commented:
@cemoody  Where is it convenient for you to put comments regarding `inv`?

delta2323 commented on cupy/cuda/cublas.py:
```diff
@@ -320,3 +320,19 @@ def sgemmBatched(handle, transa, transb, m, n, k, alpha, Aarray, lda, Barray,
+_cublas.cublasSgetrfBatched.argtypes = [Handle, _I, _P, _I, _P, _P, _I]
+
+
+def sgetrf(handle, n, A, lda, P, info, batchSize):
```
According to the naming convention of `cupy.cublas`, CuPy's method name is made by removing `cublas` and `v2` from the corresponding cublas method name, and making the first character lowercase. So this method name should be `sgetrfBatched`.

delta2323 commented on cupy/cuda/cublas.py:
```diff
@@ -320,3 +320,19 @@ def sgemmBatched(handle, transa, transb, m, n, k, alpha, Aarray, lda, Barray,
+_cublas.cublasSgetriBatched.argtypes = (Handle, _I, _P, _I, _P, _P, _I, _P, _I)
+
+
+def sgetri(handle, n, A, lda, P, C, ldac, info, batchSize):
```
Same as `sgetrf`. This method name should be `sgetriBatched`.

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_unary.py:
```diff
@@ -0,0 +1,315 @@
+        self.check_backward(x_data, y_grad, F.det)
+
+
+class TestUnaryFunctionsSimple(UnaryFunctionsTestBase, unittest.TestCase):
```
Can you use `parameterized` decorator for parameterized test cases? Example usage is [here](https://github.com/pfnet/chainer/blob/a8de5718b6197ce4882eed058e467528db2e7698/tests/cupy_tests/testing_tests/test_parameterized.py)

delta2323 commented:
Please write unit tests for `det` (not `batch_det`).

delta2323 commented:
Please add test cases where determinant is `0`, which is edge case of `backward`.

delta2323 commented:
@cemoody Do you have time to work on this PR? Since it requires some knowledge to resolve conflicts,  we will help you to fix the code.

I am sorry to make you hurry, but we will work on sending you PR if no respond is returned until the end of this week (1/8).

cemoody commented:
Hi @delta2323, looks like #608 will be merged soon, so we can focus some attention on this PR.

1) Do you want to resolve the conflicts first and then I do the fixes you have above?  
or
2) Should I try to address your concerns above first, and then you resolve the conflicts?

delta2323 commented:
Thank you for response. Then, could you try to resolve the conflict first?

delta2323 commented:
@cemoody I mistakenly think that you will work on merging the PR. I will send you PR that resolves the conflict. Sorry for waiting so much time.

delta2323 commented:
I sent you the [PR](https://github.com/cemoody/chainer/pull/3) that keeps up with the latest master. Please merge it and apply the changes I commented.

delta2323 commented:
Thank you for merging my PR! Now could you apply my comments to this PR?

cemoody commented:
Hi @delta2323, I think I've rebased master and updated everything you've requested above.

The big changes are:
- GPU and CPU modes for `batch_det` and `det` are tested
- Singular matrices are checked for. In CPU mode, NumPy raises `numpy.linalg.LingAlgError` and in GPU mode I raise `ValueError. Is this OK?
- Edge case when the determinant is zero (and thus gradient becomes NaN). 
- `det` reshapes to a scalar when relevant

cemoody commented:
The Linux and Windows builds are OK, but for some reason the Mac build is still failing. Otherwise I think the tests for this PR are passing.

delta2323 commented:
@cemoody Thank you for your work! I will check it.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+
+def _det_gpu(b):
+    # We do a batched LU decomposition on the GPU to compute
+    # and compute the determinant by multiplying the diagonal
```
Add a period at the end of the sentence.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+def _det_gpu(b):
+    # We do a batched LU decomposition on the GPU to compute
+    # and compute the determinant by multiplying the diagonal
+    # Change the shape of the array to be size=1 minibatch if necessary
```
ditto (add a period)

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+    # We do a batched LU decomposition on the GPU to compute
+    # and compute the determinant by multiplying the diagonal
+    # Change the shape of the array to be size=1 minibatch if necessary
+    # Also copy the matrix as the elments will be modified in-place
```
ditto (add a period)

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+    # Change the shape of the array to be size=1 minibatch if necessary
+    # Also copy the matrix as the elments will be modified in-place
+    a = matmul._as_batch_mat(b).copy()
+    n = int(a.shape[1])
```
Is it a must to cast to `int`?

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+    # Also copy the matrix as the elments will be modified in-place
+    a = matmul._as_batch_mat(b).copy()
+    n = int(a.shape[1])
+    n_matrices = int(a.shape[0])
```
ditto (cast to `int`).

delta2323 replied:
Replace `a.shape[0]` with `len(a)` (it is slightly faster).

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+    p = cuda.cupy.zeros((n, n_matrices), dtype='int32')
+    # Output array
+    # These arrays hold information on the execution success
+    # or if the matrix was singular
```
ditto (add a period).

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+        return 'det'
+
+    def check_type_forward(self, in_types):
+        utils.type_check.expect(in_types.size() == 1)
```
Use either `utils.type_check` or `type_check` only.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+    # The determinant is the result of the diagonal entries multiplied
+    # in each row of the minibatch
+    det = cuda.cupy.prod(a.diagonal(axis1=1, axis2=2), axis=1)
+    success = cuda.cupy.sum(info1)
```
We must check not only sum of `info1` but all of its elements should be `0`.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,108 @@
+        ~chainer.Variable: Scalar determinant of the matrix a.
+
+    """
+    shape = (1, a.data.shape[0], a.data.shape[1])
```
ditto (use `len()` instead of `shape[0]`)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,193 @@
+        x_data, y_grad = self.x.copy(), self.gy.copy()
+        gradient_check.check_backward(self.det, x_data, y_grad)
+
+    def test_expect_scalar(self):
```
Add test case for GPU?

delta2323 commented:
I am a little concerned that we just check the properties of determinant (like the transpose-invariance or the scaling law etc.). Could you add test cases where we check the result of forward propagation by comparing it with the result of calculation from the definition of determinant. Small matrices (e.g. 2x2 matrix) are enough.

cemoody commented:
Hi @delta2323, I think I've incorporated your suggestions. Let me know what you think, thanks!

delta2323 commented:
How do you think of [this comment](https://github.com/pfnet/chainer/pull/613#discussion_r52887504)?

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,217 @@
+        with self.assertRaises(ValueError):
+            gradient_check.check_backward(self.det, x_data, y_grad)
+
+    def test_answer_cpu(self):
```
How about making x randomized and calculating y with the formula det(A) = ad - bc? This change makes the tests more robust.

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,217 @@
+        gradient_check.assert_allclose(ans, y)
+
+    @attr.gpu
+    def test_answer_gpu(self):
```
ditto (make x randomized)

delta2323 commented:
Thank you for your quick response. I updated comments. Could you take a look at them?

cemoody commented:
Hi @delta2323, your idea of the randomized test is really great. I discovered that the determinant is frequently +1 or -1 of what it should be; I'm fixing now. Will update when finished!

cemoody commented:
Unfortunately, the sign of the determinant is not being calculated correctly on the GPU. I'm following the code [here](http://stackoverflow.com/questions/13393733/determinant-of-multidimensional-array), but it doesn't seem to work. If anyone knows why [this code](https://github.com/pfnet/chainer/pull/613/files#diff-4d0027dd8b2ecc0d9eae073ddfefdb26R34) doesn't work, I'd love to hear you out. In the meantime, I'll keep digging in.

delta2323 commented:
`sign` must be the determinant of permutation matrix determined by the pivot array `p`. The determinant is determined by the parity of the permutation specified by `p` (`1` for even permutation and `-1` for odd). So we must calculate the parity of `p`.

delta2323 commented:
One of the solution to calculate the parity of the permutation is to count the inversion number of `p`:
```
a = cupy.arange(len(p))
sign = ((p > p[:, None]) * (a > a[:, None])).sum() % 2
```

The code is suggested by @beam2d.

okuta commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,116 @@
+    one = cuda.cupy.float32(1.0)
+    two = cuda.cupy.float32(2.0)
+    # `sign` will be 1.0 if the pivot equals the index; -1.0 otherwise.
+    sign = one - two * (cuda.cupy.prod(p != flags, axis=0)).astype('float32')
```
How about `sign = 1.0 - 2.0 * (cuda.cupy.prod(p != flags, axis=0)).astype('float32')`?

okuta commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,116 @@
+        x, = x
+        gy, = gy
+        grad = (gy[:, None, None] * self.detx[:, None, None] *
+                numpy.linalg.inv(x.transpose([0, 2, 1])))
```
How about `numpy.linalg.inv(x.transpose(0, 2, 1)))`?

okuta commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,116 @@
+        x, = x
+        gy, = gy
+        grad = (gy[:, None, None] * self.detx[:, None, None] *
+                inv._inv_gpu(x.transpose([0, 2, 1])))
```
How about `inv._inv_gpu(x.transpose(0, 2, 1)))`?

cemoody commented:
Thanks for the suggestions @okuta. I'll correct them once we figure out the sign of the determinant problem.

Hi @delta2323, perhaps I do not understand your and @beam2d's solution. I have a minimal script below which calculates the determinant and sign, but sometimes the answer is correct and sometimes it is incorrect.

```python
from chainer import cuda
from chainer.functions.math import matmul
import numpy


def test(x):
    x = numpy.float32(x)
    answer = cuda.to_gpu(numpy.linalg.det(x))

    y = matmul._as_batch_mat(cuda.to_gpu(x).copy())
    n = y.shape[1]
    n_matrices = len(y)
    p = cuda.cupy.zeros((n, n_matrices), dtype='int32')
    info1 = cuda.cupy.zeros(n_matrices, dtype=numpy.intp)
    ap = matmul._mat_ptrs(y)
    _, lda = matmul._get_ld(y)
    cuda.cublas.sgetrfBatched(cuda.Device().cublas_handle, n, ap.data.ptr, lda,
                              p.data.ptr, info1.data.ptr, n_matrices)
    det = cuda.cupy.prod(y.diagonal(axis1=1, axis2=2), axis=1)

    # code from @beam2d and @delta2323
    a = cuda.cupy.arange(len(p))
    parity = ((p > p[:, None]) * (a > a[:, None])).sum() % 2
    sign = parity * 2.0 - 1.0

    diff = cuda.cupy.abs(det * sign - answer) / cuda.cupy.abs(answer)
    print "input"
    print x
    print "numpy answer", answer
    print "cupy answer", det * sign
    print "parity", parity
    print "sign", sign
    print "correct?", diff < 1e-4

# Note that first dimension is over minibatches
# Correct sign
x0 = [[[0.59739441,  0.83956838],
      [-0.47299126, -0.95808893]]]
test(x0)
print " "

# Incorrect sign
x1 = [[[0.85839325,  0.13828503],
       [-1.38200307, -0.26183206]]]
test(x1)
print " "

# Correct sign
x2 = [[[-0.3346124,  0.35110205],
       [0.18419617,  1.22560608]]]
test(x2)
print " "

# Try 3x3 arrays
# Correct sign
x3 = [[[0.63134915,  0.1733923, -0.21795584],
       [-1.26153779,  1.09934044, -0.5436542],
       [0.29561609,  0.82243419,  0.74502003]]]
test(x3)
print " "

# Incorrect sign
x4 = [[[-1.35632336, 0.50838292, -0.26232147],
       [1.49157524, -0.72791415, -0.633169],
       [2.02909493, -1.6098609, 0.79856992]]]
test(x4)
print " "

# A minibatch of three matrices at once
x5 = x0 + x1 + x2
test(x5)

```

Produces the following results:

```python
input
[[[ 0.59739441  0.83956838]
  [-0.47299126 -0.95808893]]]
numpy answer [-0.17524847]
cupy answer [-0.17524843]
parity 0
sign -1.0
correct [ True]

input
[[[ 0.85839325  0.13828503]
  [-1.38200307 -0.26183206]]]
numpy answer [-0.03364454]
cupy answer [ 0.03364455]
parity 0
sign -1.0
correct [False]

input
[[[-0.3346124   0.35110205]
  [ 0.18419617  1.22560608]]]
numpy answer [-0.47477466]
cupy answer [-0.47477466]
parity 0
sign -1.0
correct [ True]

input
[[[ 0.63134915  0.1733923  -0.21795584]
  [-1.26153779  1.09934044 -0.5436542 ]
  [ 0.29561609  0.82243419  0.74502003]]]
numpy answer [ 1.23145103]
cupy answer [ 1.23145092]
parity 1
sign 1.0
correct [ True]

input
[[[-1.35632336  0.50838292 -0.26232147]
  [ 1.49157524 -0.72791415 -0.633169  ]
  [ 2.02909493 -1.6098609   0.79856992]]]
numpy answer [ 1.15468156]
cupy answer [-1.15468156]
parity 1
sign 1.0
correct [False]
```

delta2323 commented:
@cemoody Sorry, I misunderstood the definition of `PivotArray` of `cublasSgetrfBatched` (see [here](http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-getrfbatched)).

Could you apply the following change?
```
--- orig.py     2016-02-18 08:59:05.357456051 +0900
+++ modified.py 2016-02-18 09:00:10.533455913 +0900
@@ -10,7 +10,7 @@
     y = matmul._as_batch_mat(cuda.to_gpu(x).copy())
     n = y.shape[1]
     n_matrices = len(y)
-    p = cuda.cupy.zeros((n, n_matrices), dtype='int32')
+    p = cuda.cupy.zeros((n_matrices, n), dtype='int32')
     info1 = cuda.cupy.zeros(n_matrices, dtype=numpy.intp)
     ap = matmul._mat_ptrs(y)
     _, lda = matmul._get_ld(y)
@@ -19,9 +19,8 @@
     det = cuda.cupy.prod(y.diagonal(axis1=1, axis2=2), axis=1)
 
     # code from @beam2d and @delta2323
-    a = cuda.cupy.arange(len(p))
-    parity = ((p > p[:, None]) * (a > a[:, None])).sum() % 2
-    sign = parity * 2.0 - 1.0
+    parity = (p.sum(axis=1) - (n * (n + 1) / 2)) % 2
+    sign = - parity * 2.0 + 1.0
 
     diff = cuda.cupy.abs(det * sign - answer) / cuda.cupy.abs(answer)
     print "input"
```

In cuBLAS document, `PivotArray` is of shape `n x batchSize`, but I found that it should be `batchSize x n`. I think it is the document bug of cuBLAS.

cemoody commented:
@delta2323 Perfect! Thanks a ton for the help! I'll update the PR with your patches!

cemoody commented:
OK! This is looking good to me. I found some more bugs in the parity logic, but worked them out. I then added another set of tests that check that the NumPy determinants and our determinants are equal for a variety of arrays of different shapes. The tests all pass!

Let me know if y'all have any more suggestions.

delta2323 commented:
@cemoody OK, I will review it again.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,113 @@
+    # the pivot array is equal to its index.
+    rng = cuda.cupy.arange(1, n + 1, dtype='int32')
+    flags = cuda.cupy.tile(rng, (n_matrices, 1))
+    parity = cuda.cupy.sum(p != flags, axis=1) % 2
```
`parity = cuda.cupy.sum(p != rng, axis=1) % 2` is enough (broadcast will occur).

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    def det_identity(self, gpu=False):
+        if self.x.ndim == 3:
+            idt = [numpy.identity(self.x.shape[1]).astype(numpy.float32)
+                   for _ in range(self.x.shape[0])]
```
Use `len(x)` instead of `x.shape[0]`

delta2323 replied:
Use `six.moves.range` instead of `range`

delta2323 replied:
`numpy.identity` has a keyword argument `dtype`. So use `numpy.identity(self.x.shape[1], dtype=numpy.float32)`

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+        if self.x.ndim == 3:
+            idt = [numpy.identity(self.x.shape[1]).astype(numpy.float32)
+                   for _ in range(self.x.shape[0])]
+            idt = numpy.array(idt).astype(numpy.float32)
```
How about using tile?
```
idt = numpy.identity(x.shape[1], dtype=numpy.float32)
numpy.tile([None], (x.shape[0], 1, 1))
```

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            idt = [numpy.identity(self.x.shape[1]).astype(numpy.float32)
+                   for _ in range(self.x.shape[0])]
+            idt = numpy.array(idt).astype(numpy.float32)
+            chk = numpy.ones(self.x.shape[0]).astype(numpy.float32)
```
`numpy.ones` has a keyword argument `dtype`.

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            idt = numpy.array(idt).astype(numpy.float32)
+            chk = numpy.ones(self.x.shape[0]).astype(numpy.float32)
+        else:
+            idt = numpy.identity(self.x.shape[1]).astype(numpy.float32)
```
ditto (use `dtype` keyword argument)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            chk = numpy.ones(self.x.shape[0]).astype(numpy.float32)
+        else:
+            idt = numpy.identity(self.x.shape[1]).astype(numpy.float32)
+            chk = numpy.ones(1).astype(numpy.float32)
```
ditto (use `dtype` keyword argument)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+
+    def det_transpose(self, gpu=False):
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
```
We do not have to copy GPU array. `self.setUp` is called and `self.x` is initialized every time we invoke each test fixture (whose name begins with `test`).

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    def det_transpose(self, gpu=False):
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
+            ct = cuda.to_gpu(self.ct.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            cx = cuda.to_gpu(self.x.copy())
+            ct = cuda.to_gpu(self.ct.copy())
+        else:
+            cx = self.x.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            ct = cuda.to_gpu(self.ct.copy())
+        else:
+            cx = self.x.copy()
+            ct = self.ct.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    def det_scaling(self, gpu=False):
+        scaling = numpy.random.randn(1).astype('float32')
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+        scaling = numpy.random.randn(1).astype('float32')
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
+            sx = cuda.to_gpu(scaling * self.x.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            cx = cuda.to_gpu(self.x.copy())
+            sx = cuda.to_gpu(scaling * self.x.copy())
+        else:
+            cx = self.x.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            sx = cuda.to_gpu(scaling * self.x.copy())
+        else:
+            cx = self.x.copy()
+            sx = scaling * self.x.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+
+    def det_product(self, gpu=False):
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    def det_product(self, gpu=False):
+        if gpu:
+            cx = cuda.to_gpu(self.x.copy())
+            cy = cuda.to_gpu(self.y.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            cx = cuda.to_gpu(self.x.copy())
+            cy = cuda.to_gpu(self.y.copy())
+        else:
+            cx = self.x.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+            cy = cuda.to_gpu(self.y.copy())
+        else:
+            cx = self.x.copy()
+            cy = self.y.copy()
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    @attr.gpu
+    @condition.retry(3)
+    def test_batch_backward_gpu(self):
+        x_data = cuda.to_gpu(self.x.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    @condition.retry(3)
+    def test_batch_backward_gpu(self):
+        x_data = cuda.to_gpu(self.x.copy())
+        y_grad = cuda.to_gpu(self.gy.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+
+    @attr.gpu
+    def test_zero_det_gpu(self):
+        x_data = cuda.to_gpu(self.x.copy())
```
ditto (no need to copy)

delta2323 commented on tests/chainer_tests/functions_tests/math_tests/test_det.py:
```diff
@@ -0,0 +1,236 @@
+    @attr.gpu
+    def test_zero_det_gpu(self):
+        x_data = cuda.to_gpu(self.x.copy())
+        y_grad = cuda.to_gpu(self.gy.copy())
```
ditto (no need to copy)

cemoody commented:
Thanks for the review @delta2323! I have now updated and addressed all of your changes.

delta2323 commented:
Thank you for the update! Now `Function.BatchDet` is LGTM. By the way, I noticed we can utilize [_det_gpu](https://github.com/pfnet/chainer/pull/613/files#diff-4d0027dd8b2ecc0d9eae073ddfefdb26R12) to implement [cupy.linalg.det](https://github.com/pfnet/chainer/blob/3a7bc835aca856cf08b23ff39ec38c70dbe735e2/cupy/linalg/norm.py#L7). We will merge `BatchDet` in the next minor release (1.7.0, 3/1). Do you think you can implement `cupy.linalg.det` until that day?

cemoody commented:
Thanks! Let's make `cupy.linalg.det` afterwards, as I don't think have enough time before this 1.7.0 to work on it.

delta2323 commented on chainer/functions/math/det.py:
```diff
@@ -0,0 +1,112 @@
+    def forward_gpu(self, x):
+        self.detx, success = _det_gpu(x[0])
+        if not success:
+            raise ValueError('Singular Matrix')
```
We do not have to raise error when the matrix is singular. We simply return 0.

delta2323 replied:
I noticed that we can determine whether we should return 0 without checking `info1`.

According to the [document](http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-getrfbatched), if some element in `info1` is non-zero, corresponding element in the diagonal (`U` in the document and `a` in your implementation) is 0. So the result of `cuda.cupy.prod(a.diagonal(axis1=1, axis2=2), axis=1)` should be 0.

delta2323 commented:
@cemoody I noticed one more point to consider. Could you check it, please?

delta2323 commented:
Also, could you consider this [suggestion](https://github.com/pfnet/chainer/pull/613/files#r44657502)?

delta2323 commented:
@cemoody Although there remains to be fixed, I will merge this PR and fix them to incorporate this feature in v1.7.0.

Thank you for your great work!

cemoody commented:
Hi @delta2323, thanks for reviewing & merging! Apologies that I didn't have time to review your later comments. :)
